staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go

```diff
@@ -333,25 +337,45 @@ func ValidateWebhookConfiguration(fldPath *field.Path, c *api.WebhookConfigurati
 	default:
 		allErrs = append(allErrs, field.NotSupported(fldPath.Child("connectionInfo", "type"), c.ConnectionInfo, []string{"InClusterConfig", "KubeConfigFile"}))
 	}
+	if len(c.MatchConditions) > 64 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("matchConditions"), c.MatchConditions, "must not exceed 64 matchConditions"))
+	}
+	for i, condition := range c.MatchConditions {
+		fldPath := fldPath.Child("matchConditions").Index(i).Child("expression")
+		if len(strings.TrimSpace(condition.Expression)) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath, ""))
+		} else {
+			allErrs = append(allErrs, validateMatchConditionsExpression(fldPath, condition.Expression)...)
+		}
+	}
+	return allErrs
+}
 
-	// TODO: Remove this check and ensure that correct validations below for MatchConditions are added
-	// for i, condition := range c.MatchConditions {
-	//	 fldPath := fldPath.Child("matchConditions").Index(i).Child("expression")
-	//	 if len(strings.TrimSpace(condition.Expression)) == 0 {
-	//	     allErrs = append(allErrs, field.Required(fldPath, ""))
-	//	 } else {
-	//		 allErrs = append(allErrs, ValidateWebhookMatchCondition(fldPath, sampleSAR, condition.Expression)...)
-	//	 }
-	// }
-	if len(c.MatchConditions) != 0 {
-		allErrs = append(allErrs, field.NotSupported(fldPath.Child("matchConditions"), c.MatchConditions, []string{}))
+func validateMatchConditionsExpression(fldPath *field.Path, expression string) field.ErrorList {
+	allErrors := field.ErrorList{}
+	compiler := authorizationcel.NewCompiler(environment.MustBaseEnvSet(environment.DefaultCompatibilityVersion()))
+	authzExpression := &authorizationcel.SubjectAccessReviewMappingCondition{
+		Expression: expression,
 	}
 
-	return allErrs
+	result, _ := compiler.CompileCELExpression(authzExpression)
+	if result.Error != nil {
+		allErrors = append(allErrors, convertCELErrorToValidationError(fldPath, authzExpression, result.Error))
+	}
+	return allErrors
 }
 
-func ValidateWebhookMatchCondition(fldPath *field.Path, sampleSAR runtime.Object, expression string) field.ErrorList {
-	allErrs := field.ErrorList{}
-	// TODO: typecheck CEL expression
-	return allErrs
+func convertCELErrorToValidationError(fldPath *field.Path, expression authorizationcel.ExpressionAccessor, err error) *field.Error {
+	var celErr *cel.Error
+	if errors.As(err, &celErr) {
+		switch celErr.Type {
+		case cel.ErrorTypeRequired:
+			return field.Required(fldPath, celErr.Detail)
+		case cel.ErrorTypeInvalid:
+			return field.Invalid(fldPath, expression.GetExpression(), celErr.Detail)
+		case cel.ErrorTypeInternal:
+			return field.InternalError(fldPath, celErr)
+		}
+	}
+	return field.InternalError(fldPath, fmt.Errorf("unsupported error type: %w", err))
```

Comment:
cel error is defined internally hence there should be no unsupported hence it might worth to go with default return internal error?

Reply:
Changing it to:
```go
        var celErr *cel.Error
	if errors.As(err, &celErr) {
		switch celErr.Type {
		case cel.ErrorTypeRequired:
			return field.Required(fldPath, celErr.Detail)
		case cel.ErrorTypeInvalid:
			return field.Invalid(fldPath, expression.GetExpression(), celErr.Detail)
		default:
			return field.InternalError(fldPath, celErr)
		}
	}
	return field.InternalError(fldPath, fmt.Errorf("error is not cel error: %w", err))
```